ensime_shared/spec/features/terrain.py

```diff
@@ -0,0 +1,115 @@
+"""
+Shared test setup such as setup/teardown hooks and fixtures.
+
+This file is loaded automatically by Lettuce, see:
+    http://lettuce.it/reference/terrain.html
+"""
+
+import os
+import shutil
+import sys
+import tempfile
+from contextlib import contextmanager
+from os import path
+
+import neovim
+from lettuce import after, before, world
+
+DEBUG = os.environ.get('DEBUG')
+LISTEN_ADDRESS = os.environ.get('NVIM_LISTEN_ADDRESS')
+
+FEATURESROOT = path.dirname(path.abspath(__file__))
+
+neovim.setup_logging()
+
+
+# TODO: Will we need a @before.each_feature that nukes .ensime_cache, etc.?
+@before.all
+def setup_test_environment():
+    world.project_path = path.join(tempfile.mkdtemp(), 'testproject')
+    world.bootstrap_projects_path = tempfile.mkdtemp()
+
+    resources = path.realpath(path.join(FEATURESROOT, '..', 'resources'))
+    shutil.copytree(path.join(resources, 'testproject'), world.project_path)
+    world.vimrc = path.join(resources, 'vimrc')
+
+
+@after.all
+def teardown_test_environment(_total):
+    if DEBUG:
+        print('-' * 80)
+        print('Project: ' + world.project_path)
+        print('Bootstraps: ' + world.bootstrap_projects_path)
+        print('Left for inspection.')
+        print('-' * 80)
+    else:
+        world.vim.quit()  # Implicitly shut server down
+        shutil.rmtree(world.project_path)
+        shutil.rmtree(world.bootstrap_projects_path)
+
+
+@before.each_scenario
+def reset_vim(_scenario):
+    # Boilerplate to set up embedded Neovim test instance, by default, allowing
+    # separate process optionally.
+    if LISTEN_ADDRESS:
+        world.vim = neovim.attach('socket', path=LISTEN_ADDRESS)
+    else:
+        embed_cmd = ['nvim', '-u', world.vimrc, '-i', 'NONE', '--embed']
+        world.vim = neovim.attach('child', argv=embed_cmd)
+
+    world.vim.input(CLEANUP_FUNC.format(vimrc=world.vimrc))
```

Comment:
> I also have a hunch that changing out the `vim` instance like this might be a source of the ENSIME server processes getting left behind after test runs—perhaps Python garbage collection steps in without the `VimLeave` getting run.

Yeah, if I do something like this it seems to take care of the orphan processes problem:

```python
# If we're re-creating the vim instance, make sure to properly fire
# VimLeave so that ENSIME process is shut down
if hasattr(world, 'vim'):
    world.vim.quit()
```

But it introduces failure that's probably a race condition. I'll see what I can work out.

One idea is, like the setting variable for where bootstrap projects live, that this might suggest new user interface could be desirable. If we had commands to stop/start/restart ENSIME server, the test suite could use them and get better determinism. Users might occasionally want them if something seems to be wonky.